Audit socket backend does not reconnect on error

Storage: consul (HA available)
Version: Vault v0.7.3
Version Sha: 0b20ae0b9b7a748d607082b1add3663a28e31b68

In experimenting with the socket audit backend I've noticed it isn't reconnecting on error.  I caused a failover to a node where the listener was down, and Vault logged the following error, which is expected:

```
2017/06/28 20:23:31.306148 [ERROR] core: failed to create audit entry: path=socket/ error=dial tcp 172.17.0.1:6500: getsockopt: connection refused
```

However, after this the socket audit backed appears to just die.  After starting the listener I see no logs, and no further socket errors are logged.  Attempting to reconfigure the backend with a HTTP PUT to /v1/sys/audit/socket has no effect, I'm unable to get it to resume audit logging.

This is undesirable behavior as the backend should resume logging when the listener comes back up, it should not fail indefinitely.

This is my complete backend config:

```json
{
  "type": "socket",
  "options": {
    "address": "172.17.0.1:6500",
    "socket_type": "tcp",
    "log_raw": "false",
    "hmac_accessor": "false",
    "format": "json"
  }
}
```

--- a/builtin/audit/socket/backend.go
+++ b/builtin/audit/socket/backend.go
@@ -72,13 +72,7 @@ func Factory(conf *audit.BackendConfig) (audit.Backend, error) {
 		logRaw = b
 	}
 
-	conn, err := net.Dial(socketType, address)
-	if err != nil {
-		return nil, err
-	}
-
 	b := &Backend{
-		connection: conn,
 		saltConfig: conf.SaltConfig,
 		saltView:   conf.SaltView,
 		formatConfig: audit.FormatterConfig{
@@ -182,6 +176,12 @@ func (b *Backend) LogResponse(auth *logical.Auth, req *logical.Request,
 }
 
 func (b *Backend) write(buf []byte) error {
+	if b.connection == nil {
+		if err := b.reconnect(); err != nil {
+			return err
+		}
+	}
+
 	err := b.connection.SetWriteDeadline(time.Now().Add(b.writeDuration))
 	if err != nil {
 		return err
@@ -196,12 +196,16 @@ func (b *Backend) write(buf []byte) error {
 }
 
 func (b *Backend) reconnect() error {
+	if b.connection != nil {
+		b.connection.Close()
+		b.connection = nil
+	}
+
 	conn, err := net.Dial(b.socketType, b.address)
 	if err != nil {
 		return err
 	}
 
-	b.connection.Close()
 	b.connection = conn
 
 	return nil

